stage0/run.go

```diff
@@ -165,20 +159,78 @@ func Prepare(cfg PrepareConfig, dir string, uuid *types.UUID) error {
 
 	cdoc, err := json.Marshal(cm)
 	if err != nil {
-		return fmt.Errorf("error marshalling pod manifest: %v", err)
+		return nil, fmt.Errorf("error marshalling pod manifest: %v", err)
 	}
+	return cdoc, nil
+}
 
-	log.Printf("Writing pod manifest")
-	fn := common.PodManifestPath(dir)
-	if err := ioutil.WriteFile(fn, cdoc, 0700); err != nil {
-		return fmt.Errorf("error writing pod manifest: %v", err)
+// validatePodManifest reads the use specified pod manifest, prepare the app images
+// and do validation for the pod manifest. If the pod manifest looks good, returns
+// the manifest as []byte.
+// TODO(yifan): More validation in the future.
+func validatePodManifest(cfg PrepareConfig, dir string) ([]byte, error) {
+	cdoc, err := ioutil.ReadFile(cfg.PodManifest)
+	if err != nil {
+		return nil, fmt.Errorf("error reading pod manifest: %v", err)
+	}
+	var cm schema.PodManifest
+	if err := json.Unmarshal(cdoc, &cm); err != nil {
+		return nil, fmt.Errorf("error unmarshaling pod manifest: %v", err)
 	}
 
-	fn = path.Join(dir, common.Stage1IDFilename)
-	if err := ioutil.WriteFile(fn, []byte(cfg.Stage1Image.String()), 0700); err != nil {
+	for _, app := range cm.Apps {
+		img := app.Image
+		am, err := prepareAppImage(cfg, img.ID, dir, cfg.UseOverlay)
+		if err != nil {
+			return nil, fmt.Errorf("error setting up image %s: %v", img, err)
+		}
+		if !app.Name.Equals(am.Name) {
+			return nil, fmt.Errorf("error: image has a different name from the app (%q vs %q)", am.Name, app.Name)
+		}
+		if am.App == nil {
+			return nil, fmt.Errorf("error: image %s has no app section", img)
+		}
+	}
+	return cdoc, nil
+}
+
+// Prepare sets up a pod based on the given config.
+func Prepare(cfg PrepareConfig, dir string, uuid *types.UUID) error {
+	var err error
+
+	if cfg.Debug {
+		log.SetOutput(os.Stderr)
+	}
+
+	log.Printf("Preparing stage1")
+	if err = prepareStage1Image(cfg, cfg.Stage1Image, dir, cfg.UseOverlay); err != nil {
+		return fmt.Errorf("error preparing stage1: %v", err)
+	}
+
+	fn := path.Join(dir, common.Stage1IDFilename)
+	if err = ioutil.WriteFile(fn, []byte(cfg.Stage1Image.String()), 0700); err != nil {
 		return fmt.Errorf("error writing stage1 ID: %v", err)
 	}
 
+	var cdoc []byte
+	if len(cfg.PodManifest) != 0 {
+		cdoc, err = validatePodManifest(cfg, dir)
+		if err != nil {
+			return fmt.Errorf("error reading pod manifest: %v", err)
```

Comment:
this is actually a case where you probably just want to return the raw error, "error reading pod manifest" is misleading and superfluous